pmd-cli/src/test/java/net/sourceforge/pmd/cli/BaseCliTest.java

```diff
@@ -23,28 +33,160 @@ abstract class BaseCliTest {
+            }
+        };
+    }
+
+    public static Matcher<String> containsStringNTimes(final int times, final String substring) {
+        return new BaseMatcher<String>() {
+
+            @Override
+            public void describeTo(Description description) {
+                description.appendText("a string containing " + times + " times the substring '" + substring + "'");
+            }
+
+            @Override
+            public boolean matches(Object o) {
+                return o instanceof String
+                    && StringUtils.countMatches((String) o, substring) == times;
+            }
+        };
+    }
+
+
+    static class CliExecutionResult {
+
+        private final ByteArrayOutputStream out;
+        private final ByteArrayOutputStream err;
+        private final CliExitCode exitCode;
+
+        CliExecutionResult(ByteArrayOutputStream out,
+                           ByteArrayOutputStream err,
+                           CliExitCode exitCode) {
+            this.out = out;
+            this.err = err;
+            this.exitCode = exitCode;
+        }
+
+        public String getOut() {
+            return out.toString();
+        }
+
+        public String getErr() {
+            return err.toString();
+        }
+
+
+        public void checkOk() {
+            assertEquals(CliExitCode.OK, exitCode);
+        }
+
+        public void checkFailed() {
+            assertEquals(CliExitCode.ERROR, exitCode);
+        }
+
+        public void checkNoErrorOutput() {
+            checkStdErr(equalTo(""));
+        }
+
+        public void checkStdOut(Matcher<? super String> matcher) {
+            assertThat(getOut(), matcher);
+        }
+
+        public void checkStdErr(Matcher<? super String> matcher) {
+            assertThat(getErr(), matcher);
+        }
+
+        /**
+         * Use this method to wrap assertions.
+         */
+        public CliExecutionResult verify(ThrowingConsumer<CliExecutionResult> actions) {
+            try {
+                actions.accept(this);
+            } catch (Exception | AssertionError e) {
+                System.out.println("TEST FAILED");
+                System.out.println("> Return code: " + exitCode);
+                System.out.println("> Standard output -------------------------");
+                System.err.println(out.toString());
+                System.err.flush();
+                System.out.println("> Standard error --------------------------");
+                System.err.println(err.toString());
+                System.err.flush();
+                System.out.println("> -----------------------------------------");
+
+                if (e instanceof Exception) {
+                    throw new AssertionFailedError("Expected no exception to be thrown", e);
+                }
+                throw (AssertionError) e;
+            }
+            return this;
+        }
+
+
+        @FunctionalInterface
+        interface ThrowingConsumer<T> {
```

Comment:
We could simply use here the JUnit5 interface, can't we? (org.junit.jupiter.api.function.ThrowingConsumer)